source/opt/passes.h

```diff
@@ -74,6 +79,138 @@ class EliminateDeadConstantPass : public Pass {
+  bool Process(ir::Module* module) override {
+    Initialize(module);
+    return ProcessImpl(module);
+  };
+
+ private:
+  // Initializes the type manager, def-use manager and get the maximal id used
+  // in the module.
+  void Initialize(ir::Module* module) {
+    type_mgr_.reset(new analysis::TypeManager(*module));
+    def_use_mgr_.reset(new analysis::DefUseManager(module));
+    for (const auto& id_def : def_use_mgr_->id_to_defs()) {
+      max_id_ = id_def.first > max_id_ ? id_def.first : max_id_;
+    }
+  };
+
+  // The real entry of processing. Iterates through the type, constants, global
+  // variable section of the given module, finds the spec constants defined
+  // with OpSpecConstantOp and OpSpecConstantComposite instructions. If the
+  // result value of those spec constants can be folded, fold them to
+  // corresponding front-end constants.
+  bool ProcessImpl(ir::Module*);
+
+  // Processes the OpSpecConstantOp instruction pointed by the given
+  // instruction iterator, folds it to normal constants if possible. Returns
+  // true if the spec constant is folded to normal constants, otherwise returns
+  // false. The instruction iterator passed in by pointer points to the
+  // OpSpecConstantOp instruction to be processed and also specifies the
+  // instruction insertion position in case new instructions need to be
+  // inserted during the process.
+  bool ProcessOpSpecConstantOp(ir::Module::inst_iterator* pos);
+
+  // Try to fold the OpSpecConstantOp CompositeExtract instruction pointed by
+  // the given instruction iterator to a normal constant defining instruction.
+  // Returns the pointer to the new constant defining instruction if succeeded.
+  // Otherwise returns nullptr.
+  ir::Instruction* DoCompositeExtract(ir::Module::inst_iterator* inst_iter_ptr);
+
+  // Try to fold the OpSpecConstantOp VectorShuffle instruction pointed by the
+  // given instruction iterator to a normal constant defining instruction.
+  // Returns the pointer to the new constant defining instruction if succeeded.
+  // Otherwise return nullptr.
+  ir::Instruction* DoVectorShuffle(ir::Module::inst_iterator* inst_iter_ptr);
+
+  // Try to fold the OpSpecConstantOp <component wise operations> instruction
+  // pointed by the given instruction iterator to a normal constant defining
+  // instruction. Returns the pointer to the new constant defining instruction
+  // if succeeded, otherwise return nullptr.
+  ir::Instruction* DoComponentWiseOperation(
+      ir::Module::inst_iterator* inst_iter_ptr);
+
+  // Creates an constant defining instruction for the given Const instance and
+  // inserts the instruction at the position specified by the given instruction
+  // iterator. Returns a pointer to the created instruction if succeeded,
+  // otherwise returns a null pointer. The instruction iterator points to the
+  // same instruction before and after the insertion.
+  ir::Instruction* BuildInstructionAndAddToModule(
```

Comment:
Would be nice to comment that this is the method that actually manages new id creation and assignment.